lib/puppet/functions/lookup.rb

```diff
@@ -1,153 +1,128 @@
+# You must provide the name of a key to look up, and can optionally provide other
+# arguments. You can combine these arguments in the following ways:
+#
+# * `lookup( <NAME>, [<VALUE TYPE>], [<MERGE BEHAVIOR>], [<DEFAULT VALUE>] )`
+# * `lookup( [<NAME>], <OPTIONS HASH> )`
+# * `lookup( as above ) |$key| { # lambda should return a default value }`
+#
+# Arguments in `[square brackets]` are optional.
+#
+# The arguments accepted by `lookup` are as follows:
+#
+# 1. `<NAME>` (string or array) --- The name of the key to look up.
+#     * This can also be an array of keys. If Puppet doesn't find anything for the
+#     first key, it will try again with the subsequent ones, only resorting to a
+#     default value if none of them succeed.
+# 2. `<VALUE TYPE>` (data type) --- A
+# [data type](http://docs.puppetlabs.com/puppet/latest/reference/lang_data_type.html)
+# that must match the retrieved value; if not, the lookup (and catalog
+# compilation) will fail. Defaults to `Data` (accepts any normal value).
+# 3. `<MERGE BEHAVIOR>` (string or hash; see **"Merge Behaviors"** below) ---
+# Whether (and how) to combine multiple values. If present, this overrides any
+# merge behavior specified in the data sources. Defaults to no value; Puppet will
+# use merge behavior from the data sources if present, and will otherwise do a
+# first-found lookup.
+# 4. `<DEFAULT VALUE>` (any normal value) --- If present, `lookup` returns this
+# when it can't find a normal value. Default values are never merged with found
+# values. Like a normal value, the default must match the value type. Defaults to
+# no value; if Puppet can't find a normal value, the lookup (and compilation) will
+# fail.
+# 5. `<OPTIONS HASH>` (hash) --- Alternate way to set the arguments above, plus
+# some less-common extra options. If you pass an options hash, you can't combine
+# it with any regular arguments (except `<NAME>`). An options hash can have the
+# following keys:
+#     * `'name'` --- Same as `<NAME>` (argument 1). You can pass this as an
+#     argument or in the hash, but not both.
+#     * `'value_type'` --- Same as `<VALUE TYPE>` (argument 2).
+#     * `'merge'` --- Same as `<MERGE BEHAVIOR>` (argument 3).
+#     * `'default_value'` --- Same as `<DEFAULT VALUE>` (argument 4).
+#     * `'default_values_hash'` (hash) --- A hash of lookup keys and default
+#     values. If Puppet can't find a normal value, it will check this hash for the
+#     requested key before giving up. You can combine this with `default_value` or
+#     a lambda, which will be used if the key isn't present in this hash. Defaults
+#     to an empty hash.
+#     * `'override'` (hash) --- A hash of lookup keys and override values. Puppet
+#     will check for the requested key in the overrides hash _first;_ if found, it
+#     returns that value as the _final_ value, ignoring merge behavior. Defaults
+#     to an empty hash.
+#
+# Finally, `lookup` can take a lambda, which should accept a single parameter.
```

Comment:
Why hedge this with "should"? In what circumstances the lambda not take a single parameter?